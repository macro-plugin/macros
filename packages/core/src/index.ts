export { createLabeledBlock } from "./block"
export { createLabeledExpr } from "./expr"
export { createSwcPlugin, transform, transformAst, transformAsync } from "./transform"
export { print, printAsync, printExpr, printExprAsync } from "./print"
export { parse, parseAsync, parseExpr, parseType } from "./parse"
export { walk, Walker } from "./walk"
export { macro, $Macro, $LitMacro, $ExprMacro, $TmplMacro, $TypeMacro, $LabeledMacro } from "./macro"
export * from "./api"
export * from "./lib"
export * from "./types"
export * from "./utils"
