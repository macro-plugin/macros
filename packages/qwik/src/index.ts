export { qwik } from "./qwik"
export { task, vtask } from "./task"
export { resource } from "./resource"
export { store } from "./store"
export { signal } from "./signal"
export { events } from "./events"
export { computed } from "./computed"
export { link, css, scoped } from "./css"
