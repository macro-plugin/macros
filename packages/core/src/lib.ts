import { Expression, Identifier } from "@swc/core"
import { createExprMacro, createTmplMacro } from "./api"
import { evalAst, evalExpr, span } from "./utils"

import { printExpr } from "./print"
import { walk } from "./walk"

export var $Eval = createExprMacro("$Eval", function (args) {
  const t = args[0].type

  if (t === "StringLiteral") return this.parseExpr(evalExpr(args[0].value))
  if (t === "TemplateLiteral") {
    const track = (name: string) => this.track(name)
    walk(args[0], {
      enter (node) {
        if (node.type === "Identifier") {
          const v = track((node as Identifier).value)
          if (v && v.value) this.replace(v.value)
        }
      }
    })

    return this.parseExpr(evalExpr(evalExpr(this.printExpr(args[0]))))
  }

  if (t === "FunctionExpression" || t === "ArrowFunctionExpression") return this.parseExpr(evalExpr(`(${this.printExpr(args[0])})(${args.slice(1).map(i => this.printExpr(i)).join(",")})`))
  return this.parseExpr(evalAst(args[0]).toString())
}, "(<T>(expr: string) => T) & (<T>(expr: T) => T) & (<F extends (...args: any) => any>(expr: F, ...args: Parameters<F>) => ReturnType<F>)")

export const printAst = (ast: object) => JSON.stringify(ast).replace(/("(start|end|ctxt)":\s*)(\d+)/g, "$10")

export var $Ast = createExprMacro("$Ast", function (args) {
  return this.parseExpr(printAst(args[0].type === "StringLiteral" ? this.parseExpr(args[0].value) : args[0]))
}, '<T>(expr: T) => import("@swc/core").Expression')

export var $Env = createExprMacro("$Env", function (args, typeArgs) {
  if (args[0].type !== "StringLiteral") throw new Error("$Env macro only support StringLiteral as input.")
  const value = process.env[args[0].value] ?? ""

  if (typeArgs && typeArgs?.[0].type === "TsKeywordType") {
    switch (typeArgs[0].kind) {
      case "boolean":
        return {
          type: "BooleanLiteral",
          value: !!value,
          span
        }
      case "number":
        return {
          type: "NumericLiteral",
          value: +value,
          span
        }
    }
  }

  return {
    type: "StringLiteral",
    span,
    value
  }
}, "<R>(key: string) => R")

export var $Stringify = createExprMacro("$Stringify", function (args, typeArgs) {
  return {
    type: "StringLiteral",
    span,
    value: typeArgs && typeArgs.length > 0 ? this.printType(typeArgs[0]) : this.printExpr(args[0])
  }
}, "((expr: any) => string) & (<T>() => string)")

export const printTmpl = (strings: string[], exprs: Expression[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + printExpr(exprs[i]).code : text
}, "")

export const printRawTmpl = (strings: string[], exprs: unknown[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + exprs[i] : text
}, "")

function walkObject (obj: object | null, onEnter: (k: string, v: unknown, parent: Record<string, any>) => unknown, grand?: Record<string, any>, grandKey?: string) {
  if (obj && typeof obj === "object") {
    const allKeys = Object.keys(obj)
    for (let i = 0; i < allKeys.length; i++) {
      const k = allKeys[i]
      const v = obj[k as keyof typeof obj]

      const r = onEnter(k, v, obj)
      if (r != null && grand) {
        grand[grandKey!] = r
        break
      }

      if (typeof v === "object") {
        walkObject(v, onEnter, obj, k)
      }
    }
  }
  return obj
}

export var $Expr = createTmplMacro("$Expr", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parseExpr(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "import(\"@swc/core\").Expression")

export var $Quote = createTmplMacro("$Quote", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parse(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "(import(\"@swc/core\").Expression)[]")
