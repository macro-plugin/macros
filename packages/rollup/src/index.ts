import { Config, createSwcPlugin, getSpanOffset, transformAsync } from "@macro-plugin/core"
import { SCRIPT_EXTENSIONS, buildTransformOptions, createFilter, createResolver, createTsLib, createTsResolver, getHasModuleSideEffects, getIdMatcher, matchScriptType, patchTsOptions, resolveTsOptions } from "@macro-plugin/shared"
import { dirname, resolve } from "path"
import { minify as swcMinify, parse as swcParse, transform as swcTransform } from "@swc/core"

import type { Plugin } from "rollup"
import { TreeshakingOptions } from "rollup"

const isMacroExternal = (id: string) => /(@macro-plugin)|@swc/.test(id)

async function rollupMacroPlugin (config?: Config): Promise<Plugin> {
  const [basicSwcOptions, macroOptions] = await buildTransformOptions(config)
  const filter = createFilter(macroOptions.include, macroOptions.exclude)
  const extensions = macroOptions.extensions || SCRIPT_EXTENSIONS

  const plugin: Plugin = {
    name: "rollupMacroPlugin",
    async transform (code, id) {
      if (!filter(id)) return null

      const target = matchScriptType(id, extensions)
      if (!target) return null

      if (macroOptions.swcTransform == null || macroOptions.swcTransform === true) {
        const { isTypeScript, isJsx, isTsx } = target

        // use swc transform typescript and jsx
        if (isTypeScript || isJsx) {
          const macroPlugin = createSwcPlugin(macroOptions, code, getSpanOffset())
          const swcOptions = patchTsOptions({
            ...basicSwcOptions,
            filename: id,
            minify: false
          }, macroOptions.tsconfig === false ? undefined : resolveTsOptions(dirname(id), macroOptions.tsconfig), isTypeScript, isTsx, isJsx)
          const program = await swcParse(code, swcOptions.jsc?.parser || { syntax: "typescript" })
          return await swcTransform(macroPlugin(program), swcOptions)
        }
      }

      // native macro transform without swc
      return await transformAsync(code, macroOptions)
    },
    options (options) {
      // ignore warning on @macro-plugin
      const oldWarn = options.onwarn
      options.onwarn = (warning, warn) => {
        if (warning.exporter && isMacroExternal(warning.exporter) && ["UNRESOLVED_IMPORT", "UNUSED_EXTERNAL_IMPORT"].includes(warning.code || "")) return
        oldWarn ? oldWarn(warning, warn) : warn(warning)
      }

      // add @macro-plugin to external
      const matchExternal = getIdMatcher(options.external)
      options.external = (source, importer, isResolved) => isMacroExternal(source) || matchExternal(source, importer, isResolved)

      // add no-side-effects support for @macro-plugin
      const oldTreeshake = options.treeshake
      const getSideEffects = getHasModuleSideEffects(typeof oldTreeshake === "object" ? oldTreeshake.moduleSideEffects : undefined)
      options.treeshake = {
        ...(typeof oldTreeshake === "object" ? oldTreeshake : {}),
        moduleSideEffects: (id, external) => isMacroExternal(id) ? false : getSideEffects(id, external),
        preset: typeof oldTreeshake === "string" ? oldTreeshake : (oldTreeshake as TreeshakingOptions | undefined)?.preset
      }
    },
    renderChunk (code) {
      if (basicSwcOptions.minify || basicSwcOptions.jsc?.minify?.mangle || basicSwcOptions.jsc?.minify?.compress) {
        return swcMinify(code, basicSwcOptions.jsc?.minify)
      }

      return null
    }
  }

  if (macroOptions.resolveTs) {
    const tslib = createTsLib()
    const tsOptions = macroOptions.tsconfig === false ? {} : resolveTsOptions(resolve("."), macroOptions.tsconfig)
    plugin.resolveId = createTsResolver(tslib, tsOptions, filter)
    plugin.load = (id) => {
      if (id === tslib.virtual) return tslib.source
    }
  } else {
    plugin.resolveId = createResolver(extensions)
  }

  return plugin
}

export default rollupMacroPlugin
