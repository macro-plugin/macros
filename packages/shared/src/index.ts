import type { Config, MacroOptions } from "@macro-plugin/core"
import { existsSync, readFile, readFileSync, rm, writeFile } from "fs"

import type { Options } from "@swc/core"
import { addHook } from "pirates"
import path from "path"
import process from "process"
import { transformSync } from "@swc/core"

export { addHook }

const nodeTargetDefaults = new Map([
  ["12", "es2018"],
  ["13", "es2019"],
  ["14", "es2020"],
  ["15", "es2021"],
  ["16", "es2021"],
  ["17", "es2022"],
])

function set (obj: any, path: string, value: any) {
  let o = obj
  const parents = path.split(".")
  const key = parents.pop() as string

  for (const prop of parents) {
    if (o[prop] == null) o[prop] = {}
    o = o[prop]
  }

  o[key] = value
}

const h = Object.prototype.hasOwnProperty
export const hasProp = <T extends object>(target: T, prop: PropertyKey) => h.call(target, prop)

export function isModule (): boolean {
  try {
    require("fs")
    return false
  } catch {}
  return true
}

export async function autoRequire<T> (id: string): Promise<Awaited<T>> {
  try {
    return require(id)
  } catch {
  }
  return (await import(id) || {}).default
}

export function transformConfig (code: string, isModule: boolean = false) {
  return transformSync(code, { module: { type: isModule ? "es6" : "commonjs" }, jsc: { parser: { syntax: "typescript" }, target: "esnext" }, swcrc: false, configFile: false }).code
}

export function hookRequire<T> (id: string): T {
  const revert = addHook(
    (code) => transformConfig(code, isModule()),
    { extensions: [".js", ".ts"] }
  )

  const r = require(id)
  revert()

  return r
}

export async function loadConfigFile (): Promise<[string | undefined, Config]> {
  const cwd = process.cwd()

  const jsConfigFile = path.join(cwd, "macros.config.js")
  if (existsSync(jsConfigFile)) {
    // {type: "module" | "commonjs"} & "macros.config.js"
    const config = await autoRequire<Config>(jsConfigFile)
    return [jsConfigFile, config]
  }

  const tsConfigFile = path.join(cwd, "macros.config.ts")
  if (existsSync(tsConfigFile)) {
    // {type: "module"} & "macros.config.ts"
    if (isModule()) {
      const jsConfigFile = path.join(cwd, ".macros.config.js")
      const code = readFileSync(tsConfigFile).toString()
      const transformed = transformConfig(code, true)
      await writeFile(jsConfigFile, transformed, (err) => {
        if (err != null) throw new Error(err.message)
      })
      const config = (await import(jsConfigFile) || {}).default
      rm(jsConfigFile, (err) => {
        if (err != null) throw new Error(err.message)
      })
      return [tsConfigFile, config]
    }
    // {type: "commonjs"} & "macros.config.ts"
    const config = hookRequire<{ default: Config }>(tsConfigFile)
    return [tsConfigFile, config?.default || {}]
  }

  return [undefined, {}]
}

/**
 * sync version of load config, not handing { type: "module" }
 */
export function loadConfigFileSync (): [string | undefined, Config] {
  const cwd = process.cwd()

  const jsConfigFile = path.join(cwd, "macros.config.js")
  if (existsSync(jsConfigFile)) {
    const config = require(jsConfigFile) as Config
    return [jsConfigFile, config]
  }

  const tsConfigFile = path.join(cwd, "macros.config.ts")
  if (existsSync(tsConfigFile)) {
    const config = hookRequire<{ default: Config }>(tsConfigFile)
    return [tsConfigFile, config?.default || {}]
  }
  return [undefined, {}]
}

export function writeDts (p: string, dts: string) {
  const emit = () => writeFile(p, dts, () => {})

  if (existsSync(p)) {
    readFile(p, (err, data) => {
      if ((err == null && dts !== data.toString()) || err) emit()
    })
  } else {
    emit()
  }
}

const SWC_OPTION_KEYS: (keyof Options)[] = ["test",
  "exclude",
  "env",
  "jsc",
  "module",
  "minify",
  "sourceMaps",
  "inlineSourcesContent",
  "script",
  "cwd",
  "caller",
  "filename",
  "root",
  "rootMode",
  "envName",
  "configFile",
  "swcrc",
  "swcrcRoots",
  "inputSourceMap",
  "sourceFileName",
  "sourceRoot",
  "plugin",
  "isModule",
  "outputPath"
]

export function extractSwcOptions<O extends object> (o: O): Options {
  const output: Options = {}
  for (const [k, v] of Object.entries(o)) {
    if (SWC_OPTION_KEYS.includes(k as keyof Options)) output[k as keyof Options] = v
  }
  return output
}

export function applySwcOptions (config: Config) {
  const swcOptions = extractSwcOptions(config)

  if (!swcOptions.jsc?.target) {
    set(
      swcOptions,
      "jsc.target",
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      nodeTargetDefaults.get(process.version.match(/v(\d+)/)![1]) || "es2018"
    )
  }

  set(swcOptions, "jsc.transform.hidden.jest", true)

  if (!swcOptions.sourceMaps) {
    set(swcOptions, "sourceMaps", "inline")
  }

  return swcOptions
}

export function applyMacroOptions (config: Config) {
  if (config.emitDts && !config.onEmitDts) {
    config.onEmitDts = (dts: string) => writeDts(path.resolve(config.dtsOutputPath || "./macros.d.ts"), dts)
  }
  return config
}

export async function buildTransformOptions (inputOptions: (Config & { experimental?: unknown }) | undefined): Promise<[Options, MacroOptions, string | undefined]> {
  const [configPath, config] = await loadConfigFile()
  const combinedOptions = { ...(inputOptions || {}), ...config }

  return [applySwcOptions(combinedOptions), applyMacroOptions(combinedOptions), configPath]
}

export function buildTransformOptionsSync (inputOptions: (Config & { experimental?: unknown }) | undefined): [Options, MacroOptions, string | undefined] {
  const [configPath, config] = loadConfigFileSync()
  const combinedOptions = { ...(inputOptions || {}), ...config }

  return [applySwcOptions(combinedOptions), applyMacroOptions(combinedOptions), configPath]
}
