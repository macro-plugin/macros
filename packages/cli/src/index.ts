/* eslint-disable no-eval */
/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable no-console */

import { MacroOptions, parse, transform } from "@macro-plugin/core"
import type { MainOptions, ShellOptions } from "./types"
import { buildTransformOptionsSync, extractFiles, extractInput, matchPattern, picomatch, writeDts } from "@macro-plugin/shared"
import { existsSync, mkdirSync, readFileSync, rmSync, writeFile, writeFileSync } from "fs"
import path, { dirname, resolve } from "path"
import { stderr, stdin } from "process"
import { transformJS, transformTS } from "../../register/src/transformer"

import { Command } from "commander"
import { createInterface } from "readline"
import pkg from "../../core/package.json"
import { spawn } from "child_process"
import watch from "node-watch"

const CHAR_COUNT = {
  "[": 0,
  "]": 0,
  "(": 0,
  ")": 0,
  "{": 0,
  "}": 0,
}

let CURRENT_CONFIG: MacroOptions

const CACHE_FILE = path.join(process.cwd(), "<repl>.ts")

const program = new Command()

program
  .name("macros")
  .description("Macro Plugin Command Line Interface.")
  .version(pkg.version)
  .passThroughOptions()
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  .option("--init [template]", "initialize macros environment")
  .option("-c, --config <file>", "use specified config file")
  .option("-e, --eval <script>", "evaluate script")
  .option("-i, --interactive", "run an interactive Macros shell")
  .option("-m, --minify", "enable/disable minification")
  .option("-o, --output <path>", "output file name or directory")
  .option("-p, --print", "print generated code")
  .action(((files: string[], options: MainOptions & { init?: boolean | string }) => {
    if (options.init) return initialize(typeof options.init === "boolean" ? "js" : options.init)

    if (files.length > 0) return files.map(i => nodeval(i, options))

    if (options.eval) {
      try {
        print(evaluate(options.eval, options))
      } catch (e) {
        console.error(e)
      }
      return
    }

    shell(options)
  }) as () => void)

program
  .command("init")
  .description("Initialize Macros environment.")
  .option("-t, --template <template>", "initialize with template, the value could be (ts, js)", "js")
  .action((options: { template?: string }) => initialize(options.template ?? "js"))

program
  .command("emit")
  .description("Emit declaration file `macros.d.ts`.")
  .action(emitTypes)

program
  .command("dev")
  .description("Transforming codes in watch mode.")
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  .option("-r, --run", "use node run the files")
  .option("-m, --minify", "enable/disable minification")
  .option("-o, --output <path>", "output file name or directory")
  .action((files: string[], options: MainOptions) => watchFiles(files, options))

program
  .command("run")
  .description("Transform and run the entry files.")
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  .option("-m, --minify", "enable/disable minification")
  .action((files: string[], options: MainOptions) => buildFiles(files, { ...options, run: true }))

program
  .command("build")
  .description("Transforming codes in production mode.")
  .argument("[files...]", "input (.js | .ts | .jsx | .tsx) files or glob patterns", [])
  // .option("--sourcemap", "output source maps for build (default: false)")
  .option("-m, --minify", "enable/disable minification")
  .option("-r, --run", "use node run the files")
  .option("-o, --output <path>", "output file name or directory")
  .option("-w, --watch", "rebuilds when modules have changed on disk")
  .action((files: string[], options: MainOptions & { watch?: boolean }) => {
    if (options.watch) return watchFiles(files, options as MainOptions)

    buildFiles(files, options as MainOptions)
  })

program
  .command("shell")
  .description("Run an interactive Macros shell.")
  .option("-a, --ast", "interactive develop ast")
  .option("-t, --transform", "interactive show transformed code")
  .option("-m, --minify", "enable/disable minification")
  .option("-s, --swc", "enable swc(typescript|jsx) transform")
  .action(options => shell(options))

program.addHelpCommand("help [command]", "Display help for command.")

program.parse()

function isExprEnd (code: string) {
  for (let i = 0; i < code.length; i++) if (code[i] in CHAR_COUNT) CHAR_COUNT[code[i] as keyof typeof CHAR_COUNT]++

  return CHAR_COUNT["("] === CHAR_COUNT[")"] && CHAR_COUNT["["] === CHAR_COUNT["]"] && CHAR_COUNT["{"] === CHAR_COUNT["}"] && !/[.?:]\s*$/.test(code)
}

function getConfig () {
  if (!CURRENT_CONFIG) CURRENT_CONFIG = buildTransformOptionsSync({})[1]
  return CURRENT_CONFIG
}

function emitTypes () {
  const config = getConfig()
  transform("", {
    ...config,
    emitDts: true,
    onEmitDts (dts) {
      writeDts(path.resolve(config.dtsOutputPath || "./macros.d.ts"), dts)
    }
  })
}

function clearChars () {
  Object.keys(CHAR_COUNT).map(i => CHAR_COUNT[i as keyof typeof CHAR_COUNT] === 0)
}

function print (result: unknown) {
  if (result && typeof result === "object" && "type" in result) {
    console.log(JSON.stringify(result, undefined, 2))
  } else if (typeof result === "string") {
    if (result !== "") console.log(JSON.stringify(result))
  } else console.log(result)
}

/** generate macros.d.ts and create a config file if not exists */
function initialize (template: string) {
  emitTypes()

  if (matchPattern(["**/macros.config.{js,cjs,mjs,ts}"]).length === 0) { writeFileSync("./macros.config." + template, 'import { defineConfig } from "@macro-plugin/core";\n\nexport default defineConfig({\n\n})\n') }
}

function evaluate (code: string, options: ShellOptions = {}) {
  return eval(transform(code, { ...getConfig(), minify: options.minify }).code)
}

function buildFiles (input: string[], options: MainOptions) {
  extractFiles(input).map(f => nodeval(f, options))
}

function watchFiles (input: string[], options: MainOptions) {
  const { files, patts } = extractInput(input)

  const pms = patts.map(i => picomatch(i))

  console.log("Start watching...")

  watch(
    "./",
    {
      recursive: true,
      filter (f) {
        if (files.includes(resolve(f))) return true
        for (const match of pms) if (match(f)) return true
        return false
      },
    },
    (evt, name) => {
      if (evt === "update") nodeval(name, options)
    }
  )
}

function nodeval (file: string, options: MainOptions = {}) {
  const macrosOutputFile = options.output || file.replace(/\.\w+$/, ".output.js")

  if (!options.run) {
    const src = readFileSync(file).toString()

    const dir = dirname(macrosOutputFile)
    if (!existsSync(dir)) mkdirSync(dir)

    writeFile(macrosOutputFile, file.endsWith(".js") ? transformJS(src) : transformTS(src, file), (err) => {
      if (err !== null) {
        console.log(`${file} → ${macrosOutputFile}`)
      }
    })

    return
  }

  const node = spawn("node", ["-r", "@macro-plugin/core/register", file], {
    env: {
      ...process.env,
      // macrosOutputFile,
    },
  })

  node.stdout.on("data", data => {
    stdin.write(data)
  })

  node.stderr.on("data", data => {
    stderr.write(data)
  })

  // node.on("close", code => {
  // if (code === 0 && !options.print) console.log(`${file} → ${macrosOutputFile}`)
  // })
}

async function shell (options: ShellOptions, history: string[] = []) {
  const readline = createInterface({
    input: process.stdin,
    output: process.stdout,
    terminal: true,
    prompt: ">>> ",
    history,
  })

  const lines: string[] = []
  let nestedLines: string[] = []

  readline.prompt()

  readline.on("line", line => {
    if (isExprEnd(line)) {
      try {
        const current = [...nestedLines, line].join("\n")
        const code = [...lines, ...nestedLines, line].join("\n")
        if (options.ast) {
          print(parse(current, { syntax: "typescript", tsx: true }).body[0])
        } else if (options.transform) {
          console.log("`" + JSON.stringify(options.swc ? transformTS(current, "<repl>.tsx") : transformJS(current)).replace(/\\"/g, "\"").slice(1, -1) + "`")
        } else {
          const result = evaluate(code, options)
          if (!/\s*(let|const|var|function|async|class|import|export|enum|while|for|throw|interface|if|switch|private|try)\s+/.test(current)) { print(result) } // preventing the use of previous return value
        }
        lines.push(current)
      } catch (e) {
        console.error(e)
      }
      clearChars()
      nestedLines = []
      readline.setPrompt(">>> ")
    } else {
      readline.setPrompt("... ")
      nestedLines.push(line)
    }
    readline.prompt()
  })

  readline.on("close", () => {
    if (existsSync(CACHE_FILE)) rmSync(CACHE_FILE)
  })
}
