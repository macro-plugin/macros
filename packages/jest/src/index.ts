import { Config, createSwcPlugin, getSpanOffset } from "@macro-plugin/core"
import { Options, parse, parseSync, version as swcVersion, transform, transformSync } from "@swc/core"
import type { TransformOptions, Transformer } from "@jest/transform"

import { buildTransformOptions } from "@macro-plugin/shared"
import { createHash } from "crypto"
import getCacheKeyFunction from "@jest/create-cache-key-function"
import { readFileSync } from "fs"
import { version } from "../package.json"

function insertInstrumentOptions (jestOptions: TransformOptions<unknown>, canInstrument: boolean, inputOptions: Options, instrumentOptions?: any) {
  const shouldInstrument = jestOptions.instrument && canInstrument

  if (!shouldInstrument) {
    return inputOptions
  }

  if (inputOptions?.jsc?.experimental?.plugins?.some((x) => x[0] === "swc-plugin-coverage-instrument")) {
    return
  }

  if (!inputOptions.jsc) {
    inputOptions.jsc = {}
  }

  if (!inputOptions.jsc.experimental) {
    inputOptions.jsc.experimental = {}
  }

  if (!Array.isArray(inputOptions.jsc.experimental.plugins)) {
    inputOptions.jsc.experimental.plugins = []
  }

  inputOptions.jsc.experimental.plugins?.push(["swc-plugin-coverage-instrument", instrumentOptions ?? {}])
}

async function createTransformer (inputOptions: Config & {
  experimental?: {
    customCoverageInstrumentation?: {
      enabled: boolean
      coverageVariable?: string,
      compact?: boolean,
      reportLogic?: boolean,
      ignoreClassMethods?: Array<string>,
      instrumentLog?: { level: string, enableTrace: boolean }
    }
  }
}): Promise<Transformer> {
  const [computedSwcOptions, macroOptions, configPath] = await buildTransformOptions(inputOptions)
  const cacheKeyFunction = getCacheKeyFunction([], [swcVersion, version, JSON.stringify(inputOptions), configPath ? readFileSync(configPath).toString() : ""])
  const { enabled: canInstrument, ...instrumentOptions } = inputOptions?.experimental?.customCoverageInstrumentation ?? {}

  return {
    canInstrument: !!canInstrument, // Tell jest we'll instrument by our own
    process (src, filename, jestOptions) {
      // Determine if we actually instrument codes if jest runs with --coverage
      insertInstrumentOptions(jestOptions, !!canInstrument, computedSwcOptions, instrumentOptions)
      const offset = getSpanOffset()
      const plugin = createSwcPlugin(macroOptions, src, offset)
      const program = parseSync(src, computedSwcOptions.jsc?.parser || { syntax: "typescript", tsx: true })

      return transformSync(plugin(program), {
        ...computedSwcOptions,
        module: {
          ...computedSwcOptions.module,
          type: (jestOptions.supportsStaticESM ? "es6" : "commonjs" as any)
        },
        filename
      })
    },
    async processAsync (src, filename, jestOptions) {
      insertInstrumentOptions(jestOptions, !!canInstrument, computedSwcOptions, instrumentOptions)
      const offset = getSpanOffset()
      const plugin = createSwcPlugin(macroOptions, src, offset)
      const program = await parse(src, computedSwcOptions.jsc?.parser || { syntax: "typescript", tsx: true })

      return await transform(plugin(program), {
        ...computedSwcOptions,
        module: {
          ...computedSwcOptions.module,
          // async transform is always ESM
          type: ("es6" as any)
        },
        filename
      })
    },

    getCacheKey (src, filename, ...rest) {
      // @ts-expect-error - type overload is confused
      const baseCacheKey = cacheKeyFunction(src, filename, ...rest)

      // @ts-expect-error - signature mismatch between Jest <27 og >=27
      const options: TransformOptions = typeof rest[0] === "string" ? rest[1] : rest[0]

      return createHash("md5")
        .update(baseCacheKey)
        .update("\0", "utf8")
        .update(JSON.stringify({ supportsStaticESM: options.supportsStaticESM }))
        .digest("hex")
    }
  }
}

export default { createTransformer }
