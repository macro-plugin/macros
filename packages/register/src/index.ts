import { transformJS, transformTS } from "./transformer"

import { addHook } from "@macro-plugin/shared"

addHook(
  transformJS,
  { extensions: [".js"] }
)

addHook(
  transformTS,
  { extensions: [".jsx", ".tsx", ".ts"] }
)
