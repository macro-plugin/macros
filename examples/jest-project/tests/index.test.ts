import { add, binaryAst } from "../src"

test("add", () => {
  expect(add(1, 4)).toBe(5)
})

test("eval macro", () => {
  expect($Eval(3 * 12 - 2)).toBe(34)
  expect($Eval("1 + 2")).toBe(3)
})

test("ast macro", () => {
  expect(binaryAst()).toMatchSnapshot()
  expect($Ast([])).toMatchSnapshot()
  expect($Ast((3).toFixed())).toMatchSnapshot()
})
