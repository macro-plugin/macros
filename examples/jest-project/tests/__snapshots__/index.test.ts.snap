// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast macro 1`] = `
{
  "left": {
    "raw": "2",
    "span": {
      "ctxt": 0,
      "end": 0,
      "start": 0,
    },
    "type": "NumericLiteral",
    "value": 2,
  },
  "operator": "*",
  "right": {
    "raw": "3",
    "span": {
      "ctxt": 0,
      "end": 0,
      "start": 0,
    },
    "type": "NumericLiteral",
    "value": 3,
  },
  "span": {
    "ctxt": 0,
    "end": 0,
    "start": 0,
  },
  "type": "BinaryExpression",
}
`;

exports[`ast macro 2`] = `
{
  "elements": [],
  "span": {
    "ctxt": 0,
    "end": 0,
    "start": 0,
  },
  "type": "ArrayExpression",
}
`;

exports[`ast macro 3`] = `
{
  "arguments": [],
  "callee": {
    "object": {
      "expression": {
        "raw": "3",
        "span": {
          "ctxt": 0,
          "end": 0,
          "start": 0,
        },
        "type": "NumericLiteral",
        "value": 3,
      },
      "span": {
        "ctxt": 0,
        "end": 0,
        "start": 0,
      },
      "type": "ParenthesisExpression",
    },
    "property": {
      "optional": false,
      "span": {
        "ctxt": 0,
        "end": 0,
        "start": 0,
      },
      "type": "Identifier",
      "value": "toFixed",
    },
    "span": {
      "ctxt": 0,
      "end": 0,
      "start": 0,
    },
    "type": "MemberExpression",
  },
  "span": {
    "ctxt": 0,
    "end": 0,
    "start": 0,
  },
  "type": "CallExpression",
  "typeArguments": null,
}
`;
